staging/src/k8s.io/apiserver/pkg/server/hooks.go

```diff
@@ -28,19 +28,19 @@ import (
 	restclient "k8s.io/client-go/rest"
 )
 
-// PostStartHookFunc is a function that is called after the server has started.
-// It must properly handle cases like:
+// HookFunc is a function that can be added to difference places in the server startup/shutdown logic.
+// If the hook is to be called after the server has started then it must properly handle cases like:
 //  1. asynchronous start in multiple API server processes
 //  2. conflicts between the different processes all trying to perform the same action
 //  3. partially complete work (API server crashes while running your hook)
 //  4. API server access **BEFORE** your hook has completed
 // Think of it like a mini-controller that is super privileged and gets to run in-process
 // If you use this feature, tag @deads2k on github who has promised to review code for anyone's PostStartHook
 // until it becomes easier to use.
-type PostStartHookFunc func(context PostStartHookContext) error
+type HookFunc func(context HookContext) error
 
-// PostStartHookContext provides information about this API server to a PostStartHookFunc
-type PostStartHookContext struct {
+// HookContext provides information about this API server to a HookFunc
+type HookContext struct {
```

Comment:
Why should these contexts be the same for start and stop? Already now it is odd to have `StopCh` in a pre-stop hook. Can we split them again?

Reply:
Originally I was thinking the StopCh would be used to co-ordinate running the pre-stop hooks in parallel, so all the hooks could run at the same time.  In practice that seems overly complicated and probably issue prone.  I'll split the contexts back out